platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesBirthRoundMigrationTests.java

```diff
@@ -0,0 +1,424 @@
+import com.swirlds.config.api.Configuration;
+import com.swirlds.config.extensions.test.fixtures.TestConfigBuilder;
+import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.event.preconsensus.PcesBirthRoundMigration;
+import com.swirlds.platform.event.preconsensus.PcesConfig_;
+import com.swirlds.platform.event.preconsensus.PcesFile;
+import com.swirlds.platform.event.preconsensus.PcesFileIterator;
+import com.swirlds.platform.event.preconsensus.PcesMutableFile;
+import com.swirlds.platform.system.BasicSoftwareVersion;
+import com.swirlds.platform.system.StaticSoftwareVersion;
+import com.swirlds.platform.test.fixtures.event.generator.StandardGraphGenerator;
+import com.swirlds.platform.test.fixtures.event.source.StandardEventSource;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class PcesBirthRoundMigrationTests {
+
+    /**
+     * Temporary directory provided by JUnit
+     */
+    @TempDir
+    private Path testDirectory;
+
+    private Path recycleBinPath;
+    private Path pcesPath;
+    private Path temporaryFilePath;
+
+    @BeforeAll
+    static void beforeAll() throws ConstructableRegistryException {
+        ConstructableRegistry.getInstance().registerConstructables("com.swirlds");
+    }
+
+    @BeforeEach
+    void beforeEach() throws IOException {
+        if (Files.exists(testDirectory)) {
+            FileUtils.deleteDirectory(testDirectory);
+        }
+
+        Files.createDirectories(testDirectory);
+        recycleBinPath = testDirectory.resolve("recycle-bin");
+        pcesPath = testDirectory.resolve("pces");
+        temporaryFilePath = testDirectory.resolve("tmp");
+
+        StaticSoftwareVersion.setSoftwareVersion(new BasicSoftwareVersion(0));
+
+        System.out.println(testDirectory); // TODO remove
+    }
+
+    @AfterEach
+    void afterEach() throws IOException {
+        if (Files.exists(testDirectory)) {
+            FileUtils.deleteDirectory(testDirectory);
+        }
+
+        StaticSoftwareVersion.reset();
+    }
+
+    private record PcesFilesWritten(@NonNull List<PcesFile> files, @NonNull List<GossipEvent> events) {}
+
+    /**
+     * Generate a bunch of PCES files in the legacy format.
+     *
+     * @param random        the random number generator
+     * @param discontinuity whether to introduce a discontinuity in the stream. If true, discontinuity is placed at the
```

Comment:
What if a discontinuity were to occur *after* the migration point?

Reply:
In theory, we might see a discontinuity after the migration point if a node reconnected and got the freeze state.

The short story is that all events in the PCES prior to the discontinuity will be ignored, and only events after the discontinuity will be eligible for migration. File cleanup ignores discontinuities/origin rounds, so that will work the same as normal.

When reading events to be migrated, we use the method `PcesFileTracker.getFileIterator(final long lowerBound, final long originRound)`, which is designed only to traverse events without crossing discontinuities. This is well unit tested in the non-migration unit tests.

Suppose a node reconnects and gets the freeze round state. There are two possible scenarios:

1) the node gets all non-ancient events through gossip prior to its peers shutting down
2) the node does not get all non-ancient events through gossip prior to its peers shutting down

Scenario 1 is easy, the migration will happen just like normal.

Scenario 2 is a bit more complex. If the network shuts down before the reconnecting node receives non-ancient events through gossip, and if the node is eventually power cycled to upgrade it (currently I think it might get stuck waiting for signatures), then it may not migrate some events. It will receive those events through gossip when it starts back up (it is impossible for this to happen to >=1/3 nodes). It will then write those events to its PCES in birth round mode. No long term harm.

Are these scenarios something you think it worthwhile to unit test?

Reply:
I think it might be worth adding to unit tests, just to cover all the edges

Reply:
Added requested test coverage.